File: compiler/src/utils/command_parser.py
"""
Generic command parsing utilities for handling space-separated token syntax.
Extracted from then_macro.py for reuse across macros.
"""

from typing import Dict, List, Optional, Any
from abc import ABC, abstractmethod


class Command(ABC):
    """Base class for command words in token parsing."""
    
    @abstractmethod
    def execute(self, tokens: List[str], pos: int, result: Dict[str, Any]) -> int:
        """Execute this command, modify result as needed, return new position."""
        pass


class SetFieldCommand(Command):
    """Handle setting a field in the result dict."""
    
    def __init__(self, field_name: str, takes_argument: bool = True):
        self.field_name = field_name
        self.takes_argument = takes_argument
    
    def execute(self, tokens: List[str], pos: int, result: Dict[str, Any]) -> int:
        if not self.takes_argument:
            result[self.field_name] = True
            return pos
        
        if pos >= len(tokens):
            return pos  # Error: no argument
        
        result[self.field_name] = tokens[pos]
        return pos + 1


class AppendToListCommand(Command):
    """Handle appending values to a list field."""
    
    def __init__(self, field_name: str):
        self.field_name = field_name
    
    def execute(self, tokens: List[str], pos: int, result: Dict[str, Any]) -> int:
        if pos >= len(tokens):
            return pos  # Error: no argument
        
        if self.field_name not in result:
            result[self.field_name] = []
        
        result[self.field_name].append(tokens[pos])
        return pos + 1


class ConsumeRestCommand(Command):
    """Handle consuming all remaining tokens into a list."""
    
    def __init__(self, field_name: str):
        self.field_name = field_name
    
    def execute(self, tokens: List[str], pos: int, result: Dict[str, Any]) -> int:
        result[self.field_name] = tokens[pos:]
        return len(tokens)


class CommandParser:
    """Generic parser for space-separated command syntax."""
    
    def __init__(self, content: str, commands: Dict[str, Command], default_fields: Dict[str, Any] = None):
        self.tokens = content.split()
        self.commands = commands
        self.default_fields = default_fields or {}
    
    def parse(self) -> Optional[Dict[str, Any]]:
        """Parse command tokens into structured components."""
        result = self.default_fields.copy()
        pos = 0
        
        # Handle the case where first token is not a command (implicit main argument)
        if self.tokens and self.tokens[0] not in self.commands:
            # First token is the main type/operation name
            result['main'] = self.tokens[0]
            pos = 1
        
        # Scan remaining tokens
        while pos < len(self.tokens):
            token = self.tokens[pos]
            
            if token in self.commands:
                command = self.commands[token]
                pos = command.execute(self.tokens, pos + 1, result)
            else:
                return None  # Error: unknown token
        
        return result


# Pre-built command sets for common patterns

def create_type_commands() -> Dict[str, Command]:
    """Create command set for type expression parsing."""
    return {
        'for': SetFieldCommand('parameter_name'),
        'is': SetFieldCommand('base_type'),
        'extends': AppendToListCommand('constraints'),
    }


def create_pipeline_commands() -> Dict[str, Command]:
    """Create command set for pipeline parsing (from original then_macro)."""
    return {
        'do': SetFieldCommand('operation'),
        'get': SetFieldCommand('operation'),  # alias for do
        'set': SetFieldCommand('operation'),  # alias for do
        'chain': ConsumeRestCommand('chain_steps'),
        'as': SetFieldCommand('bind_name'),
        'into': SetFieldCommand('assign_name'),
        'the': SetFieldCommand('source_variable'),
        'from': SetFieldCommand('source_variable'),
        'in': SetFieldCommand('source_variable'),
    }

File: compiler/src_ts/utils/command_parser.ts
// utils/command_parser.ts

// A tiny, generic parser for space-separated command syntax.
// It maps command words to small handlers that mutate a result record.

export type ResultValue = string | string[] | boolean;
export type ResultHash = Record<string, ResultValue>;

export abstract class Command {
  abstract execute(
    tokens: string[],
    pos: number,
    result: ResultHash,
  ): number;
}

// Sets a field to either `true` (flag) or to the next token.
export class SetFieldCommand extends Command {
  readonly field_name: string;
  readonly takes_argument: boolean;

  constructor(field_name: string, takes_argument: boolean = true) {
    super();
    this.field_name = field_name;
    this.takes_argument = takes_argument;
  }

  execute(tokens: string[], pos: number, result: ResultHash): number {
    if (!this.takes_argument) {
      result[this.field_name] = true;
      return pos;
    }
    if (pos >= tokens.length) {
      return pos;
    }
    result[this.field_name] = tokens[pos];
    return pos + 1;
  }
}

// Appends the next token into a list.
export class AppendToListCommand extends Command {
  readonly field_name: string;

  constructor(field_name: string) {
    super();
    this.field_name = field_name;
  }

  execute(tokens: string[], pos: number, result: ResultHash): number {
    if (pos >= tokens.length) {
      return pos;
    }
    const list = Array.isArray(result[this.field_name])
      ? (result[this.field_name] as string[])
      : [];
    list.push(tokens[pos]);
    result[this.field_name] = list;
    return pos + 1;
  }
}

// Consumes all remaining tokens into a list.
export class ConsumeRestCommand extends Command {
  readonly field_name: string;

  constructor(field_name: string) {
    super();
    this.field_name = field_name;
  }

  execute(tokens: string[], pos: number, result: ResultHash): number {
    result[this.field_name] = tokens.slice(pos);
    return tokens.length;
  }
}

// Generic parser.
export class CommandParser {
  private tokens: string[];
  private commands: Record<string, Command>;
  private default_fields: ResultHash;

  constructor(
    content: string,
    commands: Record<string, Command>,
    default_fields?: ResultHash,
  ) {
    this.tokens = content.split(/\s+/).filter((x) => x.length > 0);
    this.commands = commands;
    this.default_fields = default_fields ? { ...default_fields } : {};
  }

  parse(): ResultHash | null {
    const result: ResultHash = { ...this.default_fields };
    let pos = 0;

    // If the first token isn't a command, treat it as the main argument.
    if (this.tokens.length > 0 && !(this.tokens[0] in this.commands)) {
      result["main"] = this.tokens[0];
      pos = 1;
    }

    while (pos < this.tokens.length) {
      const token = this.tokens[pos];
      const cmd = this.commands[token];
      if (!cmd) {
        return null;
      }
      pos = cmd.execute(this.tokens, pos + 1, result);
    }

    return result;
  }
}

// Pre-built command sets.

export function create_type_commands(): Record<string, Command> {
  return {
    "for": new SetFieldCommand("parameter_name"),
    "is": new SetFieldCommand("base_type"),
    "extends": new AppendToListCommand("constraints"),
  };
}

export function create_pipeline_commands(): Record<string, Command> {
  return {
    "do": new SetFieldCommand("operation"),
    "get": new SetFieldCommand("operation"),
    "set": new SetFieldCommand("operation"),
    "chain": new ConsumeRestCommand("chain_steps"),
    "as": new SetFieldCommand("bind_name"),
    "into": new SetFieldCommand("assign_name"),
    "the": new SetFieldCommand("source_variable"),
    "from": new SetFieldCommand("source_variable"),
    "in": new SetFieldCommand("source_variable"),
  };
}
